### PR TITLE
CTX-4420: Updated qiime templates to reflect alterations made in the qiime pipeline tampate

### PR DESCRIPTION
**Pull request by VukManojlovic**
Updated all five qiime templates with the same upgrades that have been done in the bio-qiime-pipeline template, mainly:
- support for paired-end reads
- support for new Illumina dataset format
- automatic recognition of sequencing read types

### CODE REVIEW

File: bio-demultiplexing-fasta/src/demultiplexed.py [hunk 1]
```diff
@@ -35,34 +36,111 @@ def importMetadata(metadata: CustomSample, outputDir: Path) -> Path:
...
+        fastqSamples = ctx_qiime2.getFastqDPSamples(dataset)
+
+        if pairedEnd:
+            createManifestPaired(fastqSamples, inputPath)
+            sequenceType = "SampleData[PairedEndSequencesWithQuality]"
+            inputFormat = "PairedEndFastqManifestPhred33V2"
+        else:
+            createManifestSingle(fastqSamples, inputPath)
+            sequenceType = "SampleData[SequencesWithQuality]"
+            inputFormat = "SingleEndFastqManifestPhred33V2"
```
Comment (dule1322) [inline bio-demultiplexing-fasta/src/demultiplexed.py:140]: Let's extract all this logic into coretexpylib `SequenceDataset` and `SequenceSample` classes. `SequenceSample` currently supports `[fa, fq, fasta, fastq]` files with or without gz compression. Additional support for `_metadata` and `paired-end` sequences has to be implemented.

Comment (igorperic17) [reply bio-demultiplexing-fasta/src/demultiplexed.py:140]: Yes, this is a very common functionality and will be used across large amount of data jobs in the future. It definitely belongs in coretexpylib.

File: bio-demultiplexing-fasta/experiment.config [hunk 2]
```diff
@@ -38,7 +24,7 @@
         },
         {
             "name": "barcodeColumn",
-            "description": "Column from sequence metadata file which containes IDs for samples.\r\nThe values of this column are used to match sample sequences with their\r\ncorresponding barcode. Only used with multiplexed data.",
+            "description": "(If this field is empty it will be assumed that the sequencing reads are demultiplexed)\r\nColumn from sequence metadata file which containes IDs for samples.\r\nThe values of this column are used to match sample sequences with their\r\ncorresponding barcode. Only used with multiplexed data.",
```
Comment (dule1322) [inline bio-demultiplexing-fasta/experiment.config:27]: Update description.txt too

File: bio-demultiplexing-fasta/src/demultiplexed.py [hunk 3]
```diff
@@ -28,48 +29,73 @@ def importSample(manifestPath: Path, sequenceType: str, outputDir: Path) -> Path
...
         csv.writer(manifestFile, delimiter = "\t").writerow(["sample-id", "absolute-filepath"])
 
     for sample in samples:
-        for fastqPath in Path(sample.path).iterdir():
-            if fastqPath.suffix != ".fastq":
-                continue
+        sample.unzip()
+        fastqPath = sample.sequencePath
+
+        with manifestPath.open("a") as manifestFile:
```
Comment (dule1322) [inline bio-demultiplexing-fasta/src/demultiplexed.py:49]: we should not call open every single time we want to write to a file, especially in a for-loop, this slows things down, let's just open the file once

File: bio-denoise/main.py [hunk 4]
```diff
@@ -1,30 +1,75 @@
...
-def dada2DenoiseSingleSample(sample: CustomSample, denoiseAlgorithm: str, trimLeft: int, truncLen: int, outputDir: Path) -> Path:
+FORWARD_SUMMARY_NAME = "forward-seven-number-summaries.tsv"
+REVERSE_SUMMARY_NAME = "reverse-seven-number-summaries.tsv"
+
+
+def isPairedEnd(demuxSample: CustomSample):
+    demuxTemp = folder_manager.createTempFolder("demux")
+    qzaPath = list(demuxSample.path.iterdir())[0]
+
+    with ZipFile(BytesIO(qzaPath.read_bytes()), "r") as qzaFile:
```
Comment (dule1322) [inline bio-denoise/main.py:22]: replace with path

File: bio-taxonomic-analysis/experiment.config [hunk 5]
```diff
@@ -4,28 +4,28 @@
...
     "project_task": 11,
     "is_active": true,
-    "version": 5,
+    "version": 6,
     "parameters": [
         {
             "name": "dataset",
             "description": "A dataset which contains denoised fastq sequences. This\r\ndataset is output of the \"Step #2\" template. Dataset is generated\r\nby Qiime2 tool.",
-            "value": null,
+            "value": 9257,
```
Comment (dule1322) [inline bio-taxonomic-analysis/experiment.config:12]: hardcoded values

**Review by dule1322 — CHANGES_REQUESTED** [review]

File: bio-demultiplexing-fasta/src/demultiplexed.py [hunk 6]
```diff
@@ -28,48 +29,72 @@ def importSample(manifestPath: Path, sequenceType: str, outputDir: Path) -> Path
...
+            csv.writer(manifestFile, delimiter = "\t").writerow([sample.sequencePath.stem, sample.sequencePath])
+
+    return manifestPath
+
+
+def createManifestPaired(samples: List[SequenceSample], manifestPath: Path) -> Path:
+    with manifestPath.open("w") as manifestFile:
+        csv.writer(manifestFile, delimiter = "\t").writerow(["sample-id", "forward-absolute-filepath", "reverse-absolute-filepath"])
+
+    with manifestPath.open("a") as manifestFile:
```
Comment (dule1322) [inline bio-demultiplexing-fasta/src/demultiplexed.py:58]: join this open with the top one which writes the header into the file

File: bio-denoise/main.py [hunk 7]
```diff
@@ -1,30 +1,75 @@
...
 from coretex.bioinformatics import ctx_qiime2
 
 
-def dada2DenoiseSingleSample(sample: CustomSample, denoiseAlgorithm: str, trimLeft: int, truncLen: int, outputDir: Path) -> Path:
+FORWARD_SUMMARY_NAME = "forward-seven-number-summaries.tsv"
+REVERSE_SUMMARY_NAME = "reverse-seven-number-summaries.tsv"
+
+
+def isPairedEnd(demuxSample: CustomSample):
+    demuxTemp = folder_manager.createTempFolder("demux")
```
Comment (dule1322) [inline bio-denoise/main.py:19]: can we add a comment which explains how this function works and what it reads from the qiime metadata?

File: bio-denoise/main.py [hunk 8]
```diff
@@ -60,32 +105,84 @@ def featureTableTabulateSeqsSample(sample: CustomSample, outputDir: Path) -> Pat
     return visualizationPath
 
 
+def determineTruncLen(sample: CustomSample, forward: bool) -> int:
+    sample.unzip()
+
+    summariesFileName = FORWARD_SUMMARY_NAME if forward else REVERSE_SUMMARY_NAME
+    summariesTsv = list(sample.path.rglob(summariesFileName))[0]
+
+    truncLen: Optional[int] = None
```
Comment (dule1322) [inline bio-denoise/main.py:114]: move `truncLen` declaration to where the `highestScore` is

**Review by dule1322 — APPROVED** [review]

**Pull request merged**